pkg/connector/privilege.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	var rv []*v2.Resource
+	for _, privilege := range privileges {
+		ur, err := privilegeResource(&privilege)
+
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		rv = append(rv, ur)
+	}
+
+	return rv, nextPage, nil, nil
+}
+
+func (g *privilegeResourceType) Entitlements(_ context.Context, resource *v2.Resource, token *pagination.Token) ([]*v2.Entitlement, string, annotations.Annotations, error) {
+	ents := []*v2.Entitlement{
+		entitlement.NewAssignmentEntitlement(
+			resource,
+			"assign",
+			entitlement.WithDescription("Assign this privilege to a user or role"),
+			entitlement.WithDisplayName(fmt.Sprintf("Assign '%s' Privilege", resource.DisplayName)),
+			entitlement.WithGrantableTo(resourceTypeUser, resourceTypeRole),
+		),
+		entitlement.NewAssignmentEntitlement(
+			resource,
+			"has",
+			entitlement.WithDisplayName("Privilege Action "),
+			entitlement.WithDescription("Privilege actions that this privilege grants"),
+			entitlement.WithGrantableTo(resourceTypeUser, resourceTypeRole),
+			entitlement.WithAnnotation(&v2.EntitlementImmutable{}),
+		),
+	}
+
+	return ents, "", nil, nil
+}
+
+func (g *privilegeResourceType) Grants(ctx context.Context, resource *v2.Resource, token *pagination.Token) ([]*v2.Grant, string, annotations.Annotations, error) {
+	var bag pagination.Bag
+
+	err := bag.Unmarshal(token.Token)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	state := bag.Pop()
+
+	if state == nil {
+		bag.Push(pagination.PageState{
+			Token:          "",
+			ResourceTypeID: resourceTypeRole.Id,
+		})
+
+		bag.Push(pagination.PageState{
+			Token:          "",
+			ResourceTypeID: resourceTypeUser.Id,
+		})
+
+		privilege, err := g.client.GetPrivilegeById(ctx, resource.Id.Resource)
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		grants := make([]*v2.Grant, 0)
+
+		for _, statement := range privilege.Privilege.Statement {
+			for _, action := range statement.Action {
+				rsPrivilege := &v2.ResourceId{
+					ResourceType: resourceTypePrivilegeAction.Id,
+					Resource:     action,
+				}
+
+				newGrant := grant.NewGrant(resource, "has", rsPrivilege)
+				grants = append(grants, newGrant)
+			}
+		}
+
+		nextToken, err := bag.Marshal()
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		return grants, nextToken, nil, nil
+	}
+
+	grants := make([]*v2.Grant, 0)
+
+	switch state.ResourceTypeID {
+	case resourceTypeRole.Id:
+		rolesResponse, err := g.client.GetPrivilegeAssignableRoles(ctx, resource.Id.Resource, state.Token)
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		if rolesResponse.NextLink != "" {
+			bag.Push(pagination.PageState{
+				Token:          rolesResponse.NextLink,
+				ResourceTypeID: resourceTypeRole.Id,
+			})
+		}
+
+		for _, role := range rolesResponse.Roles {
+			grants = append(
+				grants,
+				grant.NewGrant(
+					resource,
+					"assign",
+					&v2.ResourceId{
+						ResourceType: resourceTypeRole.Id,
+						Resource:     role,
+					},
+				),
```

Comment:
do we need grant expansion for this? do users with the role get all the privileges?

Reply:
Yes, I added